src/main/java/au/org/ala/names/util/DwcaWriter.java

```diff
@@ -0,0 +1,413 @@
+            // write header row
+            writeHeader(writer, rowType, columns, detatched);
+        }
+
+        String[] row;
+        if (detatched) {
+            row = new String[columns.size()];
+            for (Map.Entry<Term, String> conceptTermStringEntry : rowMap.entrySet()) {
+                int column = columns.indexOf(conceptTermStringEntry.getKey());
+                row[column] = conceptTermStringEntry.getValue();
+            }
+        } else {
+            // make sure coreId is not null for extensions
+            if (coreRowType != rowType && coreId == null)
+                log.warn("Adding an {} extension record to a core without an Id! Skip this record", rowType);
+            row = new String[columns.size() + 1];
+            row[0] = coreId;
+            for (Map.Entry<Term, String> conceptTermStringEntry : rowMap.entrySet()) {
+                int column = 1 + columns.indexOf(conceptTermStringEntry.getKey());
+                row[column] = conceptTermStringEntry.getValue();
+            }
+        }
+        writer.write(row);
+    }
+
+    protected void writeHeader(TabWriter writer, Term rowType, List<Term> columns, boolean detatched) throws IOException {
+        int idx = 0;
+        String[] row = new String[columns.size() + (detatched ? 0 : 1)];
+        if (!detatched) {
+            Term idTerm;
+            if (DwcTerm.Taxon == coreRowType) {
+                idTerm = DwcTerm.taxonID;
+            } else if (DwcTerm.Occurrence == coreRowType) {
+                idTerm = DwcTerm.occurrenceID;
+            } else if (DwcTerm.Identification == coreRowType) {
+                idTerm = DwcTerm.identificationID;
+            } else if (DwcTerm.Event == coreRowType) {
+                idTerm = DwcTerm.eventID;
+            } else {
+                // default to generic dc identifier for id column
+                idTerm = DcTerm.identifier;
+            }
+            row[idx++] = idTerm.simpleName();
+        }
+
+        for (Term term : columns) {
+            row[idx++] = term.simpleName();
+        }
+        writer.write(row);
+
+        headersOut.add(rowType);
+    }
+
+
+    /**
+     * Add a single value for the current core record.
+     * Calling this method requires that #newRecord() has been called at least once,
+     * otherwise an IllegalStateException is thrown.
+     * @param term
+     * @param value
+     */
+    public void addCoreColumn(Term term, String value) {
+        // ensure we do not overwrite the coreIdTerm if one is defined
+        if (coreIdTerm != null && coreIdTerm.equals(term)) {
+            throw new IllegalStateException("You cannot add a term that was specified as coreId term");
+        }
+
+        List<Term> coreTerms = terms.get(coreRowType);
+        if (!coreTerms.contains(term)) {
+            if (useHeaders && recordNum>1){
+                throw new IllegalStateException("You cannot add new terms after the first row when headers are enabled");
```

Comment:
The check for useHeaders shouldn't be relevant. Even without the header line, this will still create a sparse file with different row lengths throughout the file.

Reply:
Agreed